### PR TITLE
feat: A new internal lookup

### PR DESCRIPTION
**Pull request by Hoverbear**
Accomplishes https://github.com/timberio/vector/issues/2842.

In this PR propose we rename the path to `Lookup` to avoid any user confusion with the `std::fs::Path` type. It also refines the RFC to use the `FromStr` and `parse::<T>()` functionalities in the standard lib.

This adds usage in `add_fields`, `remove_fields`, and `rename_fields` so far.

### CODE REVIEW

**Review by lukesteensen — APPROVED** [review]
This looks really solid! Awesome job, just a few lints to take care of and green CI.

File: src/event/lookup/mod.rs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    pub fn from_indexmap(
+        values: IndexMap<String, TomlValue>,
+    ) -> crate::Result<IndexMap<Lookup, Value>> {
+        let mut discoveries = IndexMap::new();
+        for (key, value) in values {
+            Self::recursive_step(Lookup::try_from(key)?, value, &mut discoveries)?;
+        }
+        Ok(discoveries)
+    }
+    pub fn from_toml_table(value: TomlValue) -> crate::Result<IndexMap<Lookup, Value>> {
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply src/event/lookup/mod.rs:94]: Right? It also doesn't enforce newlines!?

File: src/event/lookup/test.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    crate::test_util::trace_init();
+    let lookup = Lookup::from_str(SUFFICIENTLY_COMPLEX).unwrap();
+
+    let mut iter = lookup.iter();
+    for (index, expected) in SUFFICIENTLY_DECOMPOSED.iter().enumerate() {
+        let parsed = iter.next().expect(&format!(
+            "Expected at index {}: {:?}, got None.",
+            index, expected
+        ));
+        assert_eq!(expected, parsed, "Failed at {}", index);
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply src/event/lookup/test.rs:128]: In this case it's for the hygenic error. :)

File: src/event/lookup/mod.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        let mut peeker = self.segments.iter().peekable();
+        let mut next = peeker.next();
+        let mut maybe_next = peeker.peek();
+        while let Some(segment) = next {
+            match segment {
+                Segment::Field(_) => {
+                    match maybe_next {
+                        Some(next) if next.is_field() => write!(f, r#"{}."#, segment)?,
+                        None | Some(_) => write!(f, "{}", segment)?,
+                    };
```
Comment (fanatid) [inline src/event/lookup/mod.rs:58]: rustfmt produce extra padding compare to next match because semicolon?

File: src/event/lookup/segment.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+impl Segment {
+    pub const fn field(v: String) -> Segment {
+        Segment::Field(v)
+    }
+
+    pub fn is_field(&self) -> bool {
+        match self {
+            Segment::Field(_) => true,
+            Segment::Index(_) => false,
+        }
```
Comment (fanatid) [inline src/event/lookup/segment.rs:22]:
```suggestion
        matches!(self, Segment::Field(_))
```

**Review by fanatid — APPROVED** [review]
Not sure that I best person for review these changes, but LGTM for me.

File: src/event/lookup/mod.rs (hunk 3 above)
Comment (Hoverbear) [reply src/event/lookup/mod.rs:58]: Yeah! Can remove that!

**Pull request merged**